src/profile.c

```diff
@@ -532,11 +530,36 @@ static inline void printOutput(int fileid, int line, BOOL exec, BOOL visited)
     }
 }
 
+// Mark line as visited, and return true if the line has been previously
+// visited (executed)
+static BOOL markVisited(int fileid, UInt line)
```

Comment:
This function is not thread safe, yet is not called from a critical section either. Can't that cause problems in HPC-GAP?

So perhaps it needs a lock? Or perhaps we can make `profileState.visitedStatements` thread local -- it suffices to merge this into a single plist at the time we write out the profiles, no? (Though I guess that would break if a thread was killed before we do that...)

Reply:
(old reply!)

We only call this from one thread -- we lock profiling to a single thread (you can see this from the caller).

There could be issues to do with ownership of the datastructures here in HPC-GAP, if they are created in one thread and used in another. I'm not opposed to a fix for that, but it wouldn't shock me if various things don't work in HPC-GAP with profiling.